### PR TITLE
feat(dispatch): handle events outside zone

### PR DESCRIPTION
**Pull request by arturovt**

### CODE REVIEW

File: src/lib/core/decorators/dispatch.ts [hunk 1]
```diff
@@ -38,6 +40,39 @@ function dispatchEvent(target: any, key: string | symbol) {
...
+ * @param event - Wrapped dispatched event
+ * @param dispatch - Function that dispatches event
+ * @param zone - `NgZone` instance
+ */
+function dispatchEventIfZoneIsNotNooped(event: WrappedDispatchedEvent, dispatch: DispatchEventFactory, zone: NgZone): void {
+    const dispatchInsideZone = (event: DispatchedEvent) => zone.run(() => dispatch(event));
+
+    zone.runOutsideAngular(() => {
+        if (isObservable(event)) {
+            event.pipe(first()).subscribe((event) => dispatchInsideZone(event));
```
Comment (splincode) [inline src/lib/core/decorators/dispatch.ts]: I think you need to combine duplicate code

Comment (arturovt) [reply src/lib/core/decorators/dispatch.ts]: any ideas?

Comment (splincode) [reply src/lib/core/decorators/dispatch.ts]: duplicate code

image

Comment (arturovt) [reply src/lib/core/decorators/dispatch.ts]: i mean - any ideas how to avoid this?

Comment (splincode) [reply src/lib/core/decorators/dispatch.ts]:
```ts
function dispatchByEvent(event: Event, dispatcher: Function) {
  if (isObservable(event)) {
            event.pipe(first()).subscribe((event) => dispatcher(event));
   } else if (isPromise(event)) {
            event.then((event) => dispatcher(event));
   } else {
            dispatcher(event);
   }
}
```

```ts
dispatchByEvent(event, dispatchInsideZone);
dispatchByEvent(event, dispatchEvent(target, key));
```

**Pull request merged**